src/shims/fs.rs

```diff
@@ -722,6 +799,248 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                let off64_t_layout = this.libc_ty_layout("off64_t")?;
+                let c_ushort_layout = this.libc_ty_layout("c_ushort")?;
+                let c_uchar_layout = this.libc_ty_layout("c_uchar")?;
+
+                // If the host is a Unix system, fill in the inode number with its real value.
+                // If not, use 0 as a fallback value.
+                #[cfg(unix)]
+                let ino = std::os::unix::fs::DirEntryExt::ino(&dir_entry);
+                #[cfg(not(unix))]
+                let ino = 0;
+
+                let file_type = this.file_type_to_d_type(dir_entry.file_type())? as u128;
+
+                let imms = [
+                    immty_from_uint_checked(ino, ino64_t_layout)?, // d_ino
+                    immty_from_uint_checked(0u128, off64_t_layout)?, // d_off
+                    immty_from_uint_checked(0u128, c_ushort_layout)?, // d_reclen
+                    immty_from_uint_checked(file_type, c_uchar_layout)?, // d_type
+                ];
+                this.write_packed_immediates(entry_place, &imms)?;
+
+                let result_place = this.deref_operand(result_op)?;
+                this.write_scalar(this.read_scalar(entry_op)?, result_place.into())?;
+
+                Ok(0)
+            }
+            None => {
+                // end of stream: return 0, assign *result=NULL
+                this.write_null(this.deref_operand(result_op)?.into())?;
+                Ok(0)
+            }
+            Some(Err(e)) => match e.raw_os_error() {
+                // return positive error number on error
+                Some(error) => Ok(error),
+                None => {
+                    throw_unsup_format!("The error {} couldn't be converted to a return value", e)
+                }
+            },
+        }
+    }
+
+    fn readdir_r(
+        &mut self,
+        dirp_op: OpTy<'tcx, Tag>,
+        entry_op: OpTy<'tcx, Tag>,
+        result_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("readdir_r")?;
+
+        let dirp = this.read_scalar(dirp_op)?.to_machine_usize(this)?;
+
+        let entry_ptr = this.force_ptr(this.read_scalar(entry_op)?.not_undef()?)?;
+        let dirent_layout = this.libc_ty_layout("dirent")?;
+
+        let dir_iter = this.machine.dir_handler.streams.get_mut(&dirp).ok_or_else(|| {
+            err_unsup_format!("The DIR pointer passed to readdir_r did not come from opendir")
+        })?;
+        match dir_iter.next() {
+            Some(Ok(dir_entry)) => {
+                // Write into entry, write pointer to result, return 0 on success.
+                // The name is written with write_os_str_to_c_str, while the rest of the
+                // dirent struct is written using write_packed_Immediates.
+
+                let name_offset = dirent_layout.details.fields.offset(5);
+                let name_ptr = entry_ptr.offset(name_offset, this)?;
```

Comment:
Same concerns as above for moving definitions into the `match` and being more place-based.